services/src/main/java/org/keycloak/protocol/oidc/OIDCLoginProtocol.java

```diff
@@ -339,25 +347,59 @@ public Response sendError(AuthenticationSessionModel authSession, Error error) {
             redirectUri.addParam(OAuth2Constants.ISSUER, Urls.realmIssuer(session.getContext().getUri().getBaseUri(), realm.getName()));
         }
 
-        // Remove authenticationSession from current tab
-        new AuthenticationSessionManager(session).removeTabIdInAuthenticationSession(realm, authSession);
+        return redirectUri;
+    }
 
+    @Override
+    public ClientData getClientData(AuthenticationSessionModel authSession) {
+        return new ClientData(authSession.getRedirectUri(),
+                authSession.getClientNote(OIDCLoginProtocol.RESPONSE_TYPE_PARAM),
+                authSession.getClientNote(OIDCLoginProtocol.RESPONSE_MODE_PARAM),
+                authSession.getClientNote(OIDCLoginProtocol.STATE_PARAM));
+    }
+
+    @Override
+    public Response sendError(ClientModel client, ClientData clientData, Error error) {
+        logger.tracef("Calling sendError with clientData when authenticating with client '%s' in realm '%s'. Error: %s", client.getClientId(), realm.getName(), error);
+
+        // Should check if clientData are valid for current client
+        AuthorizationEndpointRequest req = AuthorizationEndpointRequest.fromClientData(clientData);
+        AuthorizationEndpointChecker checker = new AuthorizationEndpointChecker()
+                .event(event)
+                .client(client)
+                .realm(realm)
+                .request(req)
+                .session(session);
+        try {
+            checker.checkResponseType();
+            checker.checkRedirectUri();
+        } catch (AuthorizationEndpointChecker.AuthorizationCheckException ex) {
+            ex.throwAsErrorPageException(null);
+        }
+
+        setupResponseTypeAndMode(clientData.getResponseType(), clientData.getResponseMode());
+        OIDCRedirectUriBuilder redirectUri = buildErrorRedirectUri(clientData.getRedirectUri(), clientData.getState(), error);
```

Comment:
If somehow `client_data` is injected by modifying the content of the `state` parameter inside it,
could cause a reflected XSS attack after redirecting keycloak to:

https://localhost:8543/auth/realms/master/app/auth?error=temporarily_unavailable&error_description=Authentication+expired.+Please+try+again.&state=BADSTUFF&iss=https%3A%2F%2Flocalhost%3A8543%2Fauth%2Frealms%2Ftest

Maybe just check on the format of the `state` parameter before redirect is enough.

Reply:
Thanks for this point! But the thing is, that `state` parameter can be any opaque string. It is perfectly valid that client will use some arbitrary state parameter value like `BADSTUFF`. So I think we cannot easily verify the `state` parameter format on the server as it is not any prescribed format? According to https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-10#section-4.1.1, it is opaque value to the server.

The `state` parameter is actually supposed to be created by the client itself and then being send to the server. And server is supposed to return the same `state` back to the client after authentication (or in the error). Then it is the responsibility of the client to verify if `state` is correct. So if someone injects arbitrary `client_data` with incorrect (manually created) state, the client adapter is responsible for rejecting such authorization response.

Reply:
Thanks, just noticed by looking at the doc for error_description that the state parameter is under the control of the client so no check is needed